### PR TITLE
Extending Elixir Support

### PR DESCRIPTION
**Pull request by jalcine**
This PR relates to https://github.com/benekastah/neomake/issues/335 in an attempt to allow Neomake to use [`mix`](http://elixir-lang.org/docs/stable/mix/Mix.html) instead of vanilla Elixir to check files for problems.

### CODE REVIEW

**Comment by jalcine** [comment]
I need a bit of guidance around making the correct `errorformat`. Here's the output from neomake:

```
Neomake: Starting: mix compile.elixir /home/jacky/prjs/twch.at/api/test/models /home/jacky/prjs/twch.at/api/test/models/hashtag_test.exs
Neomake: mix stdout:
Neomake: mix stdout: == Compilation error on file web/models/hashtag.ex ==
Neomake: mix stdout:
Neomake: mix stdout done.
Neomake: mix processing 3 lines of output
Neomake: mix stdout: ** (CompileError) web/models/hashtag.ex:2: module Ecto.Schema is not loaded and could not be found
Neomake: mix stdout:     (elixir) expanding macro: Kernel.use/1
Neomake: mix stdout:     web/models/hashtag.ex:2: Twchat.Hashtag (module)
Neomake: mix stdout:     expanding macro: Twchat.Web.__using__/1
Neomake: mix stdout:     web/models/hashtag.ex:2: Twchat.Hashtag (module)
Neomake: mix stdout:     (elixir) expanding macro: Kernel.use/2
Neomake: mix stdout:     web/models/hashtag.ex:2: Twchat.Hashtag (module)
Neomake: mix stdout:
Neomake: mix stdout:
Neomake: mix stdout done.
Neomake: mix processing 9 lines of output
Neomake: mix complete with exit code 1
Neomake: File-level errors cleaned in buffer 2
```

**Comment by jalcine** [comment]
What would make this _super_ effective is the detection of a `mix.exs` or `mix.lock` file and **disabling** the vanilla Elixir support when it's found.

**Comment by alexlafroscia** [comment]
Hey, I just tried this out by adding the following to my config:

```viml
let g:neomake_elixir_mix_maker = {
      \ 'exe' : 'mix',
      \ 'args': ['compile', '--warnings-as-errors'],
      \ 'cwd': getcwd(),
      \ 'errorformat':
        \ '** %s %f:%l: %m,' .
        \ '%f:%l: warning: %m'
      \ }

let g:neomake_elixir_enabled_makers = ['mix']
```

So far it's working well; removed an error regarding a Phoenix dependency that the previous linter wasn't handling right, and still errors on things that it should.  Thanks!

**Comment by jalcine** [comment]
@alexlafroscia Glad this helped you out! Hoping we can get this support OOTB! 🍶

**Comment by alexlafroscia** [comment]
Hmm, so one weird thing that I noticed...

It seems like the entire project gets linted on save.  For example, I had an error in an `.eex` file that was caught when I saved some controller `.ex` file.  This itself isn't so bad, but because the linter doesn't _also_ run for `.eex` files, I had to manually re-run the `mix` linter from some other `.ex` file to make the error go away on the `.eex` file.  Also, the location that Neomake jumped to to in the `.eex` file did not match the error that `mix` referenced.

**Comment by jalcine** [comment]
Regarding the _very_ last part; I was about to make a lil' function that'd look up the `mix.exs` file but there's [a project that can help with that][1].

RE: not running for `.eex` files, I just gotta add a checker for that filetype `eelixir`. Thanks for that catch @alexlafroscia!

[1]: https://github.com/airblade/vim-rooter

**Comment by jalcine** [comment]
Hey, I don't work with Elixir at my day job so this hasn't gotten much love. I can take another look at this or if the maintainers feel it's necessary, can go ahead and close this. I'll be checking in a few days.

**Comment by jalcine** [comment]
Updated against latest master at 34502eb4d38aadeef.

File: autoload/neomake/makers/ft/elixir.vim [hunk 1]
```diff
@@ -1,7 +1,11 @@
 " vim: ts=4 sw=4 et
 
+function s:find_mix_file(path)
```
Comment (mhinz) [inline autoload/neomake/makers/ft/elixir.vim]: To make the linter happy, please add `abort` to the end of each `:function` line.

**Comment by mhinz** [comment]
If `s:find_mix_file()` isn't getting used yet, leave it away. Then we can quickly merge this.

File: autoload/neomake/makers/ft/elixir.vim (hunk 1 above)
Comment (jalcine) [inline autoload/neomake/makers/ft/elixir.vim]: Done!

**Comment by mhinz** [comment]
Thanks!

**Comment by slashmili** [comment]
@jalcine do you think we can run `mix compile` and save the compiled beam in another directory to avoid https://github.com/elixir-lang/elixir/issues/7528 ?

**Comment by slashmili** [comment]
I just want bring up the problem and have some discussion about it.

I'm not much  fan of temporary directory, so open for any suggestion.

cc @mhinz

**Comment by blueyed** [comment]
@slashmili 
Makes sense.  Please either create a new issue or a PR right away.

**Comment by jalcine** [comment]
Personally, I don't use this plugin anymore but I can still look into
updating my changes for shifting directories.

**Pull request merged**